### PR TITLE
Use new Freedom Metal Python-based tools

### PR DESCRIPTION
**Pull request by nategraff-sifive**
Keeping this PR as a draft until the various dependencies of this PR get rolled up into their respective master branches.

In the meantime, this is my take on converting this repository from using the freedom-devicetree-tools based flow to the new python tools. Namely, instead of doing a "fixup", this uses the [devicetree-overlay-generator](https://github.com/sifive/devicetree-overlay-generator) to create a new top-level Devicetree. This overlay generator allows us to dramatically simplify the following tools (especially the linker script generator).

I haven't yet tried to run the new wake scripts (I'm not sure how to drive this repository) but the wake type-checks so it solves my problem at least.

These changes are necessary because we're [replacing the freedom-devicetree-tools implementation of most of the tools the old wake rules used](https://github.com/sifive/freedom-devicetree-tools/pull/204).

### CODE REVIEW

File: program/freedom-metal-program.wake
Comment (nategraff-sifive) [inline program/freedom-metal-program.wake]: The overlay generator (and attributes/esdk settings generator, but it's less critical here) needs a `type` argument from the list of
- rtl
- arty
- vc707
- qemu
- hifive
- spike

This tells the overlay generator which heuristics to follow when picking which memories to use and more. Do we want to expose this as a parameter to `makeFreedomBSP`, and if so, how should that me made available for consumers of `api-generator-sifive`?

Comment (nategraff-sifive) [reply program/freedom-metal-program.wake]: Pushed a commit to pass in FPGABoardOpt if it isn't None, otherwise default to "rtl".

**Comment by nategraff-sifive** [comment]
Now that api-languages-sifive and the various new Python tools are happily in their respective masters, I'm ready for review here.

However, **Do Not Merge** until freedom-devicetree-tools is merged to `master` and bumped here, since those changes are still on their own branch.

There are a few main changes to this API that consumers of this API need to know to drive the generation of Freedom Metal programs:

### `DUT.FPGABoardOpts`

The Devicetree overlay generator and makeattributes generator need to know what "type" of target they're trying to generate for. If `FPGABoardOpts` isn't specified, they will expect to find a target for RTL simulation. To generate for Arty or VC707 FPGAs, you'll need to provide a string to `FPGABoardOpts` that has "arty" or "vc707" as a substring.

### `DUT.CustomOverlay`

If a project is using api-generator-sifive to generate Freedom Metal programs for an SoC, odds are the Devicetree overlay generator is going to look at the generated Devicetree for the target and projectile vomit because it doesn't fit any of the heuristics we use. In that case, you're going to want to use the [customOverlay API](https://github.com/sifive/devicetree-overlay-generator/blob/master/customOverlay.wake) to create an explicit description of how you want Freedom Metal to use the generated Devicetree and provide that to the `CustomOverlay` field of the `DUT` tuple.

### `TestProgramPlan`

The above changes replace the use of `Entry`, `Rom`, `Ram`, and `StdoutPath` fields in the `TestProgramPlan` tuple. Those fields are now controlled by the Devicetree Overlay Generator or otherwise supplied to the customOverlay API.

File: program/freedom-metal-program.wake
Comment (albertchen-sifive) [inline program/freedom-metal-program.wake]: when `coreDTS` and `designDTSName` are in different directories it fails to read the included dts. I think `coreFilename` should be `relative "{designDTSName}/.." coreDTS.getPathName`?

File: program/freedom-metal-program.wake
Comment (richardxia) [inline program/freedom-metal-program.wake:19]: Can you document the type of the argument to this function? It's not entirely obvious to me from looking at this, and from quickly looking at the code here, I believe that it's a data type that is defined in freedom-e-sdk.

File: program/freedom-metal-program.wake
Comment (richardxia) [inline program/freedom-metal-program.wake]: FYI, if you find yourself doing this pattern of `match` on a `Result` and just passing through the `Fail` case a lot, you can replace it with [`rmap`](https://github.com/sifive/wake/blob/2384a4e456ab7f8104e0030830f7c71f517223de/share/wake/lib/core/result.wake#L40-L42) or [`rmapPass`](https://github.com/sifive/wake/blob/2384a4e456ab7f8104e0030830f7c71f517223de/share/wake/lib/core/result.wake#L44-L48), which are like `map` and `mapFlat` for `Result` types, respectively. In addition to not having to manually write out the failure passthrough, it helps avoid deeply nesting code.

For example, in this case, you could rewrite it as:

```wake
attributesResult
| rmap (\attributes
  FreedomBSP
  linkerScript
  header.getFreedomMetalHeaderGeneratorOutputsHeader
  header.getFreedomMetalHeaderGeneratorOutputsInlineHeader
  platformHeader
  attributes
)
```

File: program/freedom-metal-program.wake
Comment (richardxia) [inline program/freedom-metal-program.wake]: This can be replaced with the more idiomatic [`getOrElse`](https://github.com/sifive/wake/blob/2384a4e456ab7f8104e0030830f7c71f517223de/share/wake/lib/core/option.wake#L30):

```wake
def type = dut.getDUTFPGABoardOpt | getOrElse "rtl"
```

**Review by richardxia — COMMENTED** [review]
I think we'll want to make sure it's clear how consumers of the old API can upgrade to the new API. Do `block-pio-sifive` or any of our internal repositories need to be updated to match these changes?

**Comment by nategraff-sifive** [comment]
Bumping this, I've added edits based on review. Does anyone else have feedback or can we get this merged soon?

File: wit-manifest.json
Comment (nategraff-sifive) [inline wit-manifest.json]: Just to be clear, this commit hash points at an unmerged branch. Once approved I can roll up a series of PRs, bump this hash, and then this PR can be merged.

**Review by albertchen-sifive — APPROVED** [review]
LGTM

**Review by richardxia — APPROVED** [review]

**Comment by nategraff-sifive** [comment]
Ready for merge!

**Comment by richardxia** [comment]
Hey, sorry for the delay; I've been trying to run this commit against the latest master of https://github.com/sifive/block-pio-sifive, just so I can see what the upgrading steps will be like, but I realized that I am not super familiar with what exactly needs to be changed. I'm getting the following error, which I believe implies that I need to set the test memory/boot memory via the new CustomOverlay method, but I need to track down exactly where that is:

```
$ wake runSim pioDUT
'<ivycache>' ivycache
WARNING: Could not parse DTS JSON: _root_.sifive.skeleton.DefaultConfig_sifive.skeleton.WithSimUART_sifive.skeleton.WithTestFinisher_sifive.blocks.pio.WithpioTop.json file not found
python3.7 -B devicetree-overlay-generator/generate_overlay.py -o build/freedom-metal/pioDUT/pioDUT-design.dts -t rtl --rename-include pioDUT.dts build/api-generator-sifive/pioDUT/firrtl/pioDUT.dts
devicetree-overlay-generator/generate_overlay.py: Unable to determine test bench reset vector

===== TEST RESULTS =====
- test: demo
  msg: FAILED: freedom-metal-dut-program-compiler: failed to compile binary

Fail "Some tests failed"
```

Note that I think this has nothing to do with your PR, and more to do with how I need to understand how https://github.com/sifive/soc-testsocket-sifive interacts with api-generator-sifive and block-pio-sifive.

**Comment by nategraff-sifive** [comment]
soc-testsocket-sifive uses a RocketChipDUTPlan. Adding support for these tools just means adding a few fields to the `RocketChipDUTPlan` tuple and then mapping those through when it gets converted to a `DUT`. Pushed the necessary changes here, and [soc-testsocket-sifive/build.wake](https://github.com/sifive/soc-testsocket-sifive/blob/master/build.wake) can be modified in the same way that @albertchen-sifive has already shown works internally.

**Review by albertchen-sifive — APPROVED** [review]
tested that new changes work with block-pio-sifive locally

**Review by richardxia — APPROVED** [review]
LGTM!

**Comment by nategraff-sifive** [comment]
Ready for someone to hit the merge button (I don't have privilege on this repository).

**Pull request merged**
